### PR TITLE
fix correlated ms not exported (sirius)

### PR DESCRIPTION
**Pull request by SteffenHeu**

### CODE REVIEW

File: src/main/java/io/github/mzmine/modules/io/export_features_sirius/SiriusExportTask.java [hunk 1]
```diff
@@ -128,6 +127,32 @@ protected SiriusExportTask(ParameterSet parameters, @NotNull Instant moduleCallD
...
+
+  private static void putFeatureFieldsIntoEntry(Feature f, SpectralLibraryEntry entry) {
+    int charge = 1;
+    PolarityType polarity = f.getRepresentativeScan().getPolarity();
+    if (f.getRow().getRowCharge() != null) {
+      charge = f.getRow().getRowCharge();
+    } else if (f.getMostIntenseFragmentScan().getMsMsInfo() instanceof DDAMsMsInfo dda) {
+      charge = dda.getPrecursorCharge() != null ? dda.getPrecursorCharge() : charge;
+    }
+    charge = charge == 0 ? 1 : charge;
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/modules/io/export_features_sirius/SiriusExportTask.java:148]: Math.max(1, charge)
maybe?

File: src/main/java/io/github/mzmine/modules/io/export_features_sirius/SiriusExportTask.java [hunk 2]
```diff
@@ -497,14 +487,39 @@ private MassSpectrum generateCorrelationSpectrum(@NotNull FeatureListRow row,
...
-          dps.add(new SimpleDataPoint(ip.getMzValue(i), ip.getIntensityValue(i)));
+        dps.add(new SimpleDataPoint(ip.getMzValue(i), ip.getIntensityValue(i)));
+      }
+    }
+  }
+
+  private void removeDuplicateDataPoints(List<DataPoint> dp, MZTolerance tolerance) {
+    for (int i = 0; i < dp.size() - 1; i++) {
+      if (mzTol.checkWithinTolerance(dp.get(i).getMZ(), dp.get(i + 1).getMZ())) {
+        if (dp.get(i) instanceof AnnotatedDataPoint) {
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/modules/io/export_features_sirius/SiriusExportTask.java:501]: documentation would be great. on why 
What if both are AnnotatedDataPoint

File: src/main/java/io/github/mzmine/modules/io/export_features_sirius/SiriusExportTask.java [hunk 3]
```diff
@@ -497,14 +487,39 @@ private MassSpectrum generateCorrelationSpectrum(@NotNull FeatureListRow row,
...
-        if (mzTol.checkWithinTolerance(feature.getMZ(), ip.getMzValue(i)) && exportMolecularIon) {
-          dps.add(new SimpleDataPoint(ip.getMzValue(i), ip.getIntensityValue(i)));
+        dps.add(new SimpleDataPoint(ip.getMzValue(i), ip.getIntensityValue(i)));
+      }
+    }
+  }
+
+  private void removeDuplicateDataPoints(List<DataPoint> dp, MZTolerance tolerance) {
+    for (int i = 0; i < dp.size() - 1; i++) {
+      if (mzTol.checkWithinTolerance(dp.get(i).getMZ(), dp.get(i + 1).getMZ())) {
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/modules/io/export_features_sirius/SiriusExportTask.java:500]: this should be the case but this only works if the data points are sorted by mz... 
So maybe add this to the var name sortedDps

File: src/main/java/io/github/mzmine/modules/io/export_features_sirius/SiriusExportTask.java [hunk 4]
```diff
@@ -128,6 +127,32 @@ protected SiriusExportTask(ParameterSet parameters, @NotNull Instant moduleCallD
...
+    int charge = 1;
+    PolarityType polarity = f.getRepresentativeScan().getPolarity();
+    if (f.getRow().getRowCharge() != null) {
+      charge = f.getRow().getRowCharge();
+    } else if (f.getMostIntenseFragmentScan().getMsMsInfo() instanceof DDAMsMsInfo dda) {
+      charge = dda.getPrecursorCharge() != null ? dda.getPrecursorCharge() : charge;
+    }
+    charge = charge == 0 ? 1 : charge;
+
+    entry.putIfNotNull(DBEntryField.FEATURE_ID, f.getRow().getID());
```
[the remark this answers is not in the document]
Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/modules/io/export_features_sirius/SiriusExportTask.java:150]: what the hell

[the remark this answers is not in the document]
Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/modules/io/export_features_sirius/SiriusExportTask.java:150]: but sirius does not need it so i wouldnt do it

File: src/main/java/io/github/mzmine/modules/io/export_features_sirius/SiriusExportTask.java [hunk 5]
```diff
@@ -497,14 +487,39 @@ private MassSpectrum generateCorrelationSpectrum(@NotNull FeatureListRow row,
...
-        // make sure to not export the molecular ion twice. Mass might change a bit due to smoothing
-        if (mzTol.checkWithinTolerance(feature.getMZ(), ip.getMzValue(i)) && exportMolecularIon) {
-          dps.add(new SimpleDataPoint(ip.getMzValue(i), ip.getIntensityValue(i)));
+        dps.add(new SimpleDataPoint(ip.getMzValue(i), ip.getIntensityValue(i)));
+      }
+    }
+  }
+
+  private void removeDuplicateDataPoints(List<DataPoint> dp, MZTolerance tolerance) {
+    for (int i = 0; i < dp.size() - 1; i++) {
```
Comment (robinschmid) [inline src/main/java/io/github/mzmine/modules/io/export_features_sirius/SiriusExportTask.java]: only increment i if no element was removed. if an element was removed keep i.
Better go from high to low to not miss a step. You can keep decrementing then.
````java
for(int i=size-2; i>=0; i--)
````
Or filter into a separate list

**Review by robinschmid — APPROVED** [review]
Great thanks for the changes

**Pull request merged**